glib-macros/src/lib.rs

```diff
@@ -265,6 +266,101 @@ pub fn clone(item: TokenStream) -> TokenStream {
     clone::clone_inner(item)
 }
 
+/// The same as [`gclosure_local!`](crate::gclosure_local) but uses [`Closure::new`] as a
+/// constructor.
+///
+/// [`Closure::new`]: https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/closure/struct.Closure.html#method.new
+#[proc_macro]
+#[proc_macro_error]
+pub fn gclosure(item: TokenStream) -> TokenStream {
+    gclosure::gclosure_inner(item, "new")
+}
+
+/// Macro for creating a [`Closure`] object. This is a wrapper around
+/// [`Closure::new_local`] that type checks its arguments at runtime.
+///
+/// A `Closure` takes [`Value`] objects as inputs and output. This macro will
+/// automatically convert the inputs to Rust types when invoking its callback, and then will
+/// convert the output back to a `Value`. All inputs must implement the
+/// [`FromValue`] trait, and outputs must either implement the
+/// [`ToValue`] trait or be the unit type `()`.
+///
+/// Similarly to `clone`, this macro can be useful in combination with signal handlers to reduce
+/// boilerplate when passing references. The `@weak`, `@weak-allow-none` and `@strong` captures are
+/// supported. There is one major difference compared to `clone`: references captured with `@weak`
+/// will always cause the closure to become invalidated when the object is destroyed. When a
+/// closure connected to a signal becomes invalidated, the signal handler will automatically be
+/// disconnected. Internally, this is accomplished using
+/// [`Object::watch_closure`] on each of the weak references.
```

Comment:
This also brings a big different to the `clone!` macro: That one can also support weak references for `Arc` and `Rc`, for example. But this only works for `Object` now. Which is fine, but should probably be documented :)

Reply:
Is there a case to be made for supporting the same `@weak` behavior as `clone!` has? It would be possible to do that and then move the `watch_closure` behavior to a new directive like `@weak-obj` or something like that. But that would also mean supporting `@default-return` and `@default-panic` as well.

Reply:
I think it's fine for now, just needs to be documented. If someone wants this behaviour we can add it later, and for now the `weak-none` variant can be used together with an `unwrap()` or so.

Reply:
On thinking about this, I feel it should be changed from `@weak` here, to `@weak-obj` or `@watch` or something like that. The semantics are totally different from the use of `@weak` in `clone!` and it seems that giving it the same name will cause more confusion.

Reply:
This is a very good point @jf2048. The naming would be confusing between the two.

Reply:
Good idea!

Reply:
I changed it to `@watch` and updated the docs to say that only `Object` can be passed there, also added a note specifically mentioning that `@weak-allow-none` and `@strong` still take `Arc` and `Rc`.